ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoredTypedCache.java

```diff
@@ -0,0 +1,119 @@
+package com.outbrain.ob1k.cache.metrics;
+
+import java.util.Map;
+import com.google.common.base.Preconditions;
+import com.outbrain.ob1k.cache.EntryMapper;
+import com.outbrain.ob1k.cache.TypedCache;
+import com.outbrain.ob1k.concurrent.ComposableFuture;
+import com.outbrain.ob1k.concurrent.Consumer;
+import com.outbrain.ob1k.concurrent.Try;
+import com.outbrain.swinfra.metrics.api.MetricFactory;
+import com.outbrain.swinfra.metrics.api.Timer;
+
+/**
+ * A decorator adding metric monitoring to TypedCache.
+ *
+ * @author hunchback
+ */
+public class MonitoredTypedCache<K, V> implements TypedCache<K, V> {
+  protected final TypedCache<K, V> inner;
+
+  private final AsyncGetMetrics getAsyncMetrics;
+  private final AsyncBulkGetMetrics getBulkAsyncMetric;
+  private final AsyncOperationMetrics setAsyncMetrics;
+  private final AsyncOperationMetrics setBulkAsyncMetrics;
+  private final AsyncOperationMetrics deleteAsyncMetrics;
+
+  public MonitoredTypedCache(final TypedCache<K, V> inner, String cacheName, final MetricFactory metricFactory) {
+    Preconditions.checkNotNull(inner);
+    this.inner = inner;
+    Preconditions.checkNotNull(metricFactory);
+    final String component = inner.getClass().getSimpleName() + "." + cacheName;
+
+    getAsyncMetrics = new AsyncGetMetrics(metricFactory, component + ".getAsync");
+    getBulkAsyncMetric = new AsyncBulkGetMetrics(metricFactory, component + ".getBulkAsync");
+    setAsyncMetrics = new AsyncOperationMetrics(metricFactory, component + ".setAsync");
+    setBulkAsyncMetrics = new AsyncOperationMetrics(metricFactory, component + ".setBulkAsync");
+    deleteAsyncMetrics = new AsyncOperationMetrics(metricFactory, component + ".deleteAsync");
+  }
+
+  @Override
+  public ComposableFuture<V> getAsync(final K key) {
+    final Timer.Context timerCtx = getAsyncMetrics.start();
+    ComposableFuture<V> future = inner.getAsync(key);
+    future.consume(new Consumer<V>() {
+      @Override
+      public void consume(Try<V> res) {
+       getAsyncMetrics.finished(res, timerCtx);
```

Comment:
The only context I could think of where passing the timerContext into the method makes sense is if you want to avoid creating consumers, but you didn't do that...